### PR TITLE
feat: slashing

### PR DESCRIPTION
**Pull request by 0xShaito**
# 🤖 Linear

Closes GRT-149

### CODE REVIEW

File: src/contracts/HorizonAccountingExtension.sol [hunk 1]
```diff
@@ -176,22 +194,30 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
...
+    // Release the winning pledges to the user
+    uint256 _totalPledged = _result.bondSize * _numberOfPledges;
+    _unbond(_pledger, _totalPledged);
 
-    // pledgerClaimed[_requestId][_pledger] = true;
-    // balanceOf[_pledger] += _claimAmount;
+    // Check the balance in the contract
+    // If not enough balance, slash some users to get enough balance
+    uint256 _balance = GRT.balanceOf(address(this));
+    // How many iterations should we do?
```
Comment (0xShaito) [inline src/contracts/HorizonAccountingExtension.sol:204]: Add a TODO here

File: src/contracts/HorizonAccountingExtension.sol [hunk 2]
```diff
@@ -176,22 +194,30 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
...
+
+    // Send the user the amount they won by participating in the dispute
+    GRT.safeTransfer(_pledger, _claimAmount - _totalPledged);
 
     emit EscalationRewardClaimed({
       _requestId: _requestId,
       _disputeId: _disputeId,
       _pledger: _pledger,
-      _amount: _claimAmount
+      _amount: _claimAmount - _totalPledged
```
Comment (0xShaito) [inline src/contracts/HorizonAccountingExtension.sol:220]: Can we have both parameters? reward and released amount

File: src/contracts/HorizonAccountingExtension.sol [hunk 3]
```diff
@@ -228,17 +254,25 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
     // TODO: To pay the users first we need to have slashed the losing users
```
Comment (0xShaito) [inline src/contracts/HorizonAccountingExtension.sol:254]: Remove?

File: src/contracts/HorizonAccountingExtension.sol [hunk 4]
```diff
@@ -290,19 +324,73 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
...
+    uint256 _usersToSlash,
+    uint256 _maxUsersToCheck
+  ) internal returns (uint256 _slashedAmount) {
+    EnumerableSet.AddressSet storage _users = _pledgers[_disputeId];
+
+    uint256 _slashedUsers;
+    address _user;
+    uint256 _slashAmount;
+
+    _maxUsersToCheck = _maxUsersToCheck > _users.length() ? _users.length() : _maxUsersToCheck;
```
Comment (0xShaito) [inline src/contracts/HorizonAccountingExtension.sol:369]: Save the length in a variable before

File: test/unit/HorizonAccountingExtension.t.sol [hunk 5]
```diff
@@ -411,22 +460,36 @@ contract HorizonAccountingExtension_Unit_OnSettleBondEscalation is HorizonAccoun
...
 
     assertEq(_requestIdSaved, _mockRequestId);
     assertEq(_amountPerPledgerSaved, _amountPerPledger);
-    assertEq(address(_bondEscalationModule), address(this));
+    assertEq(address(_bondEscalationModule), address(bondEscalationModule));
+    assertEq(_bondSizeSaved, _bondSize);
   }
 }
 
+// TODO: Add tests for the claimEscalationReward function with slash. As the tests is too long, we have an issue with the via ir.
```
Comment (0xShaito) [inline test/unit/HorizonAccountingExtension.t.sol:477]: Add them in a different file maybe?

File: test/unit/HorizonAccountingExtension.t.sol [hunk 6]
```diff
@@ -994,4 +1155,196 @@ contract HorizonAccountingExtension_Unit_Release is HorizonAccountingExtension_U
 
     horizonAccountingExtension.release(_bonder, _requestId, _amount);
   }
+
+  function test_revertIfInsufficientBondedTokens(
+    address _bonder,
+    bytes32 _requestId,
+    uint256 _amount
+  ) public happyPath(_bonder, _requestId, _amount) {
+    horizonAccountingExtension.setBondedTokensForTest(_bonder, 0);
```
Comment (0xShaito) [inline test/unit/HorizonAccountingExtension.t.sol:1164]: Set amount lower than amount

File: test/unit/HorizonAccountingExtension.t.sol [hunk 7]
```diff
@@ -994,4 +1155,196 @@ contract HorizonAccountingExtension_Unit_Release is HorizonAccountingExtension_U
...
+  EnumerableSet.AddressSet internal _cleanPledgers;
+
+  modifier happyPath(
+    uint256 _usersToSlash,
+    uint256 _maxUsersToCheck,
+    address[] memory _users,
+    uint256 _pledgesForDispute,
+    uint256 _bondSize
+  ) {
+    vm.assume(_usersToSlash > 0 && _usersToSlash < type(uint16).max);
```
Comment (0xShaito) [inline test/unit/HorizonAccountingExtension.t.sol:1203]: Can't we jsut have _usersToSlash be a uint16 variable in the fuzzing? (applies for all)

File: src/contracts/HorizonAccountingExtension.sol (hunk 3 above)
Comment (ashitakah) [reply src/contracts/HorizonAccountingExtension.sol:254]: 👍

File: src/contracts/HorizonAccountingExtension.sol (hunk 2 above)
Comment (ashitakah) [reply src/contracts/HorizonAccountingExtension.sol:220]: 👍

File: src/contracts/HorizonAccountingExtension.sol (hunk 1 above)
Comment (ashitakah) [reply src/contracts/HorizonAccountingExtension.sol:204]: 👍

File: src/contracts/HorizonAccountingExtension.sol [hunk 8]
```diff
@@ -14,16 +14,12 @@ import {
 
 import {Validator} from '@defi-wonderland/prophet-core/solidity/contracts/Validator.sol';
 
+import 'forge-std/console.sol';
```
Comment (0xShaito) [inline src/contracts/HorizonAccountingExtension.sol:17]: remove

File: src/contracts/HorizonAccountingExtension.sol [hunk 9]
```diff
@@ -251,26 +271,14 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
...
-    // B) Stake + Provision to the receiver
-    // Slash the payer
-    HORIZON_STAKING.slash(
-      _payer,
-      _amount,
-      _amount, // TODO: How do we manage the max verifier cut?
-      // What if it's not 100%?
-      _receiver // Maybe we should pay the receiver directly
-    );
+    // Increase the receiver bond
```
Comment (0xShaito) [inline src/contracts/HorizonAccountingExtension.sol:280]: This does not increase the receivers bond. We are sending the tokens directly to their wallet

Comment (0xShaito) [reply src/contracts/HorizonAccountingExtension.sol:280]: Lets keep the todo and see if we should stake more for them or just send them the tokens in a future task

File: src/contracts/HorizonAccountingExtension.sol [hunk 10]
```diff
@@ -393,6 +392,32 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
...
+   * @param _result The escalation result.
+   * @param _status The dispute status.
+   */
+  function _calculateSlashAmount(
+    address _pledger,
+    EscalationResult memory _result,
+    IOracle.DisputeStatus _status
+  ) internal view returns (uint256 _slashAmount) {
+    bytes32 _requestId = _result.requestId;
+    if (pledgerClaimed[_requestId][_pledger]) revert HorizonAccountingExtension_AlreadyClaimed();
```
Comment (0xShaito) [inline src/contracts/HorizonAccountingExtension.sol:407]: Why is this here? What if a user pledged 2 for and 2 against? We still need to slash him for the 2 that lost and if they claim first this would get bricked right?

File: src/contracts/HorizonAccountingExtension.sol [hunk 11]
```diff
@@ -164,34 +190,54 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
...
+    uint256 _rewardAmount;
 
     if (_status == IOracle.DisputeStatus.NoResolution) {
       _numberOfPledges = _result.bondEscalationModule.pledgesForDispute(_requestId, _pledger)
         + _result.bondEscalationModule.pledgesAgainstDispute(_requestId, _pledger);
+
+      _pledgeAmount = _result.bondSize * _numberOfPledges;
+      _claimAmount = _amountPerPledger * _numberOfPledges;
+
+      _unbond(_pledger, _pledgeAmount);
```
Comment (0xShaito) [inline src/contracts/HorizonAccountingExtension.sol:204]: The unbond can be done outside of the if for both

File: src/contracts/HorizonAccountingExtension.sol [hunk 12]
```diff
@@ -164,34 +190,54 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
...
+    uint256 _pledgeAmount;
+    uint256 _claimAmount;
+    uint256 _rewardAmount;
 
     if (_status == IOracle.DisputeStatus.NoResolution) {
       _numberOfPledges = _result.bondEscalationModule.pledgesForDispute(_requestId, _pledger)
         + _result.bondEscalationModule.pledgesAgainstDispute(_requestId, _pledger);
+
+      _pledgeAmount = _result.bondSize * _numberOfPledges;
+      _claimAmount = _amountPerPledger * _numberOfPledges;
```
Comment (0xJabberwock) [inline src/contracts/HorizonAccountingExtension.sol:202]: It's worth commenting that `_pledgeAmount` and `_claimAmount` must be equal.

File: src/contracts/HorizonAccountingExtension.sol [hunk 13]
```diff
@@ -290,19 +332,90 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
     // 1) return the funds to the disputer in case there is no resolution
     // 2) release the initial dispute bond if the disputer wins
 
-    // if (bondedAmountOf[_bonder][_token][_requestId] < _amount) revert AccountingExtension_InsufficientFunds();
+    // Release the bond amount for the request for the user
+    bondedForRequest[_bonder][_requestId] -= _amount;
```
Comment (0xJabberwock) [inline src/contracts/HorizonAccountingExtension.sol:336]: Without the explicit custom error, the reversion is still produced but at the expense of worsening DX.

File: src/contracts/HorizonAccountingExtension.sol [hunk 14]
```diff
@@ -225,20 +271,16 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
...
-
-    // if (bondedAmountOf[_payer][_token][_requestId] < _amount) {
-    //   revert AccountingExtension_InsufficientFunds();
-    // }
+    // Discount the payer bondedForRequest
+    bondedForRequest[_payer][_requestId] -= _amount;
 
-    // balanceOf[_receiver][_token] += _amount;
+    // Discout the payer totalBonded
+    totalBonded[_payer] -= _amount;
```
Comment (0xJabberwock) [inline src/contracts/HorizonAccountingExtension.sol:278]: Call `_unbond()`.

File: src/contracts/HorizonAccountingExtension.sol [hunk 15]
```diff
@@ -313,6 +426,7 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
   function _bond(address _bonder, uint256 _amount) internal {
     IHorizonStaking.Provision memory _provisionData = HORIZON_STAKING.getProvision(_bonder, address(this));
 
+    if (_provisionData.maxVerifierCut != MAX_VERIFIER_CUT) revert HorizonAccountingExtension_InvalidMaxVerifierCut();
     if (_provisionData.thawingPeriod != MIN_THAWING_PERIOD) revert HorizonAccountingExtension_InvalidThawingPeriod();
```
Comment (0xJabberwock) [inline src/contracts/HorizonAccountingExtension.sol:430]: Shouldn't the `!=` be a `<`?

**Review by 0xJabberwock — CHANGES_REQUESTED** [review]

File: test/unit/HorizonAccountingExtension.t.sol [hunk 16]
```diff
@@ -100,7 +121,7 @@ contract HorizonAccountingExtension_Unit_BaseTest is Test, Helpers {
 
     user = makeAddr('User');
 
-    horizonAccountingExtension = new HorizonAccountingExtensionForTest(horizonStaking, oracle, grt, 30 days);
+    horizonAccountingExtension = new HorizonAccountingExtensionForTest(horizonStaking, oracle, grt, 30 days, 1_000_000);
```
Comment (0xJabberwock) [inline test/unit/HorizonAccountingExtension.t.sol:124]: I'd put the magic numbers in constants, also because they appear many times.

File: src/contracts/HorizonAccountingExtension.sol (hunk 8 above)
Comment (ashitakah) [reply src/contracts/HorizonAccountingExtension.sol:17]: 👍

File: src/contracts/HorizonAccountingExtension.sol (hunk 9 above)
Comment (ashitakah) [reply src/contracts/HorizonAccountingExtension.sol:280]: 👍

File: src/contracts/HorizonAccountingExtension.sol (hunk 11 above)
Comment (ashitakah) [reply src/contracts/HorizonAccountingExtension.sol:204]: 👍

File: src/contracts/HorizonAccountingExtension.sol (hunk 15 above)
Comment (ashitakah) [reply src/contracts/HorizonAccountingExtension.sol:430]: As far I know has to be equal.

File: test/unit/HorizonAccountingExtension.t.sol (hunk 16 above)
Comment (ashitakah) [reply test/unit/HorizonAccountingExtension.t.sol:124]: 👍

File: src/contracts/HorizonAccountingExtension.sol (hunk 14 above)
Comment (ashitakah) [reply src/contracts/HorizonAccountingExtension.sol:278]: 👍

File: src/contracts/HorizonAccountingExtension.sol (hunk 13 above)
Comment (ashitakah) [reply src/contracts/HorizonAccountingExtension.sol:336]: I think that is not necessary but im open to put it @0xShaito

File: src/contracts/HorizonAccountingExtension.sol (hunk 12 above)
Comment (ashitakah) [reply src/contracts/HorizonAccountingExtension.sol:202]: 👍

File: src/contracts/HorizonAccountingExtension.sol (hunk 10 above)
Comment (ashitakah) [reply src/contracts/HorizonAccountingExtension.sol:407]: I think I follow you, the user would then not be able to claim what was earned if this is set to true in the slash, the point is, how do we know that a user has already been slashed and not have to do it again?

Comment (ashitakah) [reply src/contracts/HorizonAccountingExtension.sol]: 👍  solved talking is discord

File: src/contracts/HorizonAccountingExtension.sol (hunk 9 above)
Comment (0xShaito) [reply src/contracts/HorizonAccountingExtension.sol:275]: Change the comment of increase the receiver bond + add a comment to see what we should do if pay the user or increase their stake

File: src/contracts/HorizonAccountingExtension.sol (hunk 4 above)
Comment (0xShaito) [reply src/contracts/HorizonAccountingExtension.sol:369]: Bump

File: src/contracts/HorizonAccountingExtension.sol (hunk 15 above)
Comment (0xShaito) [reply src/contracts/HorizonAccountingExtension.sol:424]: Let's change it to <. It doesn't really give us any issues if it's bigger

**Review by 0xJabberwock — APPROVED** [review]

**Review by 0xShaito — COMMENTED** [review]
Approved! LGTM!
